src/app/qgisapp.cpp

```diff
@@ -7761,6 +7761,9 @@ void QgisApp::changeDataSource( QgsMapLayer *layer )
           }
         }
       }
+
+      if ( QgsProject *project = layer->project() )
+        project->setDirty( true );
```

Comment:
You can safely use QgsProject::instance() here -- everything in app assumes this.

```suggestion
      QgsProject::instance()->setDirty( true );
```